algosdk/future/atomic_transaction_composer.py

```diff
@@ -0,0 +1,496 @@
+                        "expected TransactionWithSigner as method argument, but received: {}".format(
+                            method_args[i]
+                        )
+                    )
+                txn_list.append(method_args[i])
+            elif len(app_args) > 14:
+                # Pack the remaining values as a tuple
+                additional_types.append(arg.type)
+                additional_args.append(method_args[i])
+            else:
+                encoded_arg = arg.type.encode(method_args[i])
+                app_args.append(encoded_arg)
+
+        if additional_args:
+            remainder_args = TupleType(additional_types).encode(
+                additional_args
+            )
+            app_args.append(remainder_args)
+
+        # Create a method call transaction
+        method_txn = transaction.ApplicationCallTxn(
+            sender=sender,
+            sp=sp,
+            index=app_id,
+            on_complete=on_complete,
+            app_args=app_args,
+            note=note,
+            lease=lease,
+            rekey_to=rekey_to,
+        )
+        txn_with_signer = TransactionWithSigner(method_txn, signer)
+        txn_list.append(txn_with_signer)
+
+        self.txn_list += txn_list
+        self.method_dict[len(self.txn_list) - 1] = method_call
+
+    def build_group(self):
+        """
+        Finalize the transaction group and returns the finalized transactions with signers.
+        The composer's status will be at least BUILT after executing this method.
+
+        Returns:
+            list[TransactionWithSigner]: list of transactions with signers
+        """
+        if self.status >= AtomicTransactionComposerStatus.BUILT:
+            return self.txn_list
+        if not self.txn_list:
+            raise error.AtomicTransactionComposerError(
+                "no transactions to build for AtomicTransactionComposer"
+            )
+
+        # Get group transaction id
+        group_txns = [t.txn for t in self.txn_list]
+        group_id = transaction.calculate_group_id(group_txns)
+        for t in self.txn_list:
+            t.txn.group = group_id
+            self.tx_ids.append(t.txn.get_txid())
+
+        self.status = AtomicTransactionComposerStatus.BUILT
+        return self.txn_list
+
+    def gather_signatures(self):
+        """
+        Obtain signatures for each transaction in this group. If signatures have already been obtained,
+        this method will return cached versions of the signatures.
+        The composer's status will be at least SIGNED after executing this method.
+        An error will be thrown if signing any of the transactions fails.
+
+        Returns:
+            list[SignedTransactions]: list of signed transactions
+        """
+        if self.status >= AtomicTransactionComposerStatus.SIGNED:
+            # Return cached versions of the signatures
+            return self.signed_txns
+
+        stxn_list = [None] * len(self.txn_list)
```

Comment:
The approach I took to gathering signatures was:
1. Loop through the transaction list and make a map that maps a unique signer to a list of transaction indexes.
2. Loop through the map and call the `sign` method on the transaction list and indexes.
3. After receiving the signed transactions for the indexes, assign it to a fixed array of signed transactions to return.
Open to discussing more elegant solutions.